### ISSUE TITLE
[BUG] `env.recv()` outputs empty numpy arrays

### ISSUE THREAD

**Issue by vwxyzjn** [sail-sg/envpool#259]
## Describe the bug

When `num_envs=batch_size`, if `env.recv()` is called twice, it returns empty numpy arrays.

## To Reproduce

```python
import envpool

env = envpool.make("Breakout-v5", "gym", num_envs=8, batch_size=8)
env.async_reset()
for _ in range(3):
    obs, _, _, _ = env.recv()
    print(obs.shape)

print("having a smaller batch_size")
env = envpool.make("Breakout-v5", "gym", num_envs=8, batch_size=4)
env.async_reset()
for _ in range(100):
    obs, _, _, _ = env.recv()
    print(obs.shape)
```
```
(8, 4, 84, 84)
(0, 4, 84, 84) # do not block but output empty arrays
(0, 4, 84, 84)
having a smaller batch_size
(4, 4, 84, 84)
(4, 4, 84, 84) # blocks
```

## Expected behavior

`env.recv()` should block in the first case.

## System info

Describe the characteristic of your environment:
 * Describe how the library was installed (pip, source, ...)
 * Python version
 * Versions of any other relevant libraries

```python
import envpool, numpy, sys
print(envpool.__version__, numpy.__version__, sys.version, sys.platform)
```
```
0.8.2 1.23.5 3.8.10 (default, Mar 13 2023, 10:26:41) 
[GCC 9.4.0] linux
```

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://envpool.readthedocs.io/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by Trinkle23897** [comment]
You shouldn’t call recv twice. It’s a bug from your code

**Comment by vwxyzjn** [comment]
I was trying to optimize the IMPALA code (https://github.com/vwxyzjn/cleanba/pull/6) using `tfds` (tensorflow dataset) so that maybe it can somehow asynchronously load `env.recv()`. I gave it a quick try with PPO, where it has a `batch_size != num_envs` and is not impacted by this empty array issue. However, it appears `tfds` does not really help. Closing this issue now.

<img width="825" alt="image" src="https://user-images.githubusercontent.com/5555347/231205882-2c4603c6-2363-4f75-bba0-fd66dcbfc892.png">
